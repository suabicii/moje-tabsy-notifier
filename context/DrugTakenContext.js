import React, {createContext, useContext} from "react";

const DrugTakenContext = createContext([]);

export const useDrugTakenContext = () => useContext(DrugTakenContext);

export default DrugTakenContext;