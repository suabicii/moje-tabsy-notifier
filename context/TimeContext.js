import React, {createContext} from "react";

export const TimeContext = createContext(undefined);