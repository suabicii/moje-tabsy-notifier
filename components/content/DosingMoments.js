import React, {useContext, useState} from "react";
import {useDrugTakenContext} from "../../context/DrugTakenContext";
import dayjs from "dayjs";
import DosingMoment from "./DosingMoment";
import {TimeContext} from "../../context/TimeContext";

const isSameOrAfter = require('dayjs/plugin/isSameOrAfter');
dayjs.extend(isSameOrAfter);

function DosingMoments({drugName, content}) {
    const {drugTakenChecker} = useDrugTakenContext();
    const {currentTime} = useContext(TimeContext);
    const [dosingMomentsToShow, setDosingMomentsToShow] = useState(content);

    const result = [];

    const handleSetDosingMomentsToShow = name => {
        setDosingMomentsToShow(
            dosingMomentsToShow.filter(dosingMoment => dosingMoment[0] !== name)
        );
    };

    for (const [key, value] of dosingMomentsToShow) {
        const viewKey = `${drugName}_${key}`;

        if (drugTakenChecker.find(string => string === viewKey)) {
            continue;
        }

        const [hour, minutes] = value.split(':');
        const dosingDateTime = dayjs().hour(hour).minute(minutes);
        const confirmationBtnDisabled = !currentTime.isSameOrAfter(dosingDateTime);

        result.push(
            <DosingMoment
                key={viewKey}
                name={key}
                drugName={drugName}
                time={value}
                id={viewKey}
                disabled={confirmationBtnDisabled}
                handleSetDosingMomentsToShow={handleSetDosingMomentsToShow}
            />
        );
    }

    return result;
}

export default DosingMoments;