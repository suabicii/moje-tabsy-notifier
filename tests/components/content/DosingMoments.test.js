import React from "react";
import renderer from "react-test-renderer";
import DosingMoments from "../../../components/content/DosingMoments";
import {drugList} from "../../fixtures/drugList";
import {fireEvent, render, screen} from "@testing-library/react-native";
import {DrugTakenContext} from "../../../context/DrugTakenContext";
import {TimeContext} from "../../../context/TimeContext";
import dayjs from "dayjs";
import MockDate from "mockdate";

let currentTime;

beforeAll(() => {
    MockDate.set('2020-01-01');
    currentTime = dayjs();
});

const drug = drugList[0];
const handleConfirmDose = jest.fn();
const dosingMomentsContent = Object.entries(drug.dosingMoments);
const drugTakenChecker = [];
const setDrugTakenChecker = jest.fn();
const setCurrentTime = jest.fn();
const mockUseContext = jest.fn().mockImplementation(() => ({drugTakenChecker, setDrugTakenChecker}));

React.useContext = mockUseContext;

function WrappedComponent() {
    return (
        <TimeContext.Provider value={{currentTime, setCurrentTime}}>
            <DrugTakenContext.Provider value={{drugTakenChecker, setDrugTakenChecker}}>
                <DosingMoments
                    drugName={drug.name}
                    content={dosingMomentsContent}
                    handleConfirmDose={handleConfirmDose}
                />
            </DrugTakenContext.Provider>
        </TimeContext.Provider>
    );
}

it('should correctly render DosingMoments component', () => {
    const tree = renderer.create(<WrappedComponent/>).toJSON();
    expect(tree).toMatchSnapshot();
});

it('should shorten dosing moment list after pressing the button to confirm the dose', () => {
    render(<WrappedComponent/>);

    fireEvent.press(screen.getByTestId('Xanax_hour1'));

    expect(screen.queryByTestId('Xanax_hour1')).toBeNull();
});