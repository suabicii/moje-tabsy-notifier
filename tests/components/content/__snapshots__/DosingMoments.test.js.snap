// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render DosingMoments component 1`] = `
[
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
        "marginBottom": 10,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#000000",
            "fontFamily": "System",
            "fontWeight": "400",
          },
          {
            "textAlign": "left",
          },
          [
            {
              "textAlign": "left",
            },
            {
              "color": "rgba(0, 0, 0, 0.87)",
              "fontFamily": "System",
              "fontWeight": "400",
              "writingDirection": "ltr",
            },
            [
              {
                "fontSize": 14,
                "letterSpacing": 0.25,
                "lineHeight": 20,
                "marginVertical": 2,
              },
              {
                "color": "rgb(28, 28, 30)",
                "fontSize": 16,
                "textAlign": "center",
              },
            ],
          ],
        ]
      }
    >
      – 
      07:00
    </Text>
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#6cc3d5",
          "borderColor": "transparent",
          "borderRadius": 4,
          "borderStyle": "solid",
          "borderWidth": 0,
          "color": "#e8e8e8",
          "elevation": 2,
          "marginLeft": 5,
          "minWidth": "auto",
          "shadowColor": "#000000",
          "shadowOffset": {
            "height": 0.75,
            "width": 0,
          },
          "shadowOpacity": 0.24,
          "shadowRadius": 1.5,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "disabled": false,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "overflow": "hidden",
            },
            {
              "borderRadius": 4,
            },
          ]
        }
        testID="Xanax_hour1"
      >
        <View
          style={
            [
              {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
              },
              {
                "flexDirection": "row-reverse",
              },
            ]
          }
        >
          <View
            style={
              {
                "marginLeft": -4,
                "marginRight": 12,
              }
            }
          >
            <Text />
          </View>
          <Text
            numberOfLines={1}
            selectable={false}
            style={
              [
                {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                {
                  "textAlign": "left",
                },
                [
                  {
                    "letterSpacing": 1,
                    "marginHorizontal": 16,
                    "marginVertical": 9,
                    "textAlign": "center",
                  },
                  {
                    "marginHorizontal": 8,
                  },
                  {
                    "textTransform": "uppercase",
                  },
                  {
                    "color": "#ffffff",
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  {
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  undefined,
                ],
              ]
            }
          >
            Już zażyłem/-am
          </Text>
        </View>
      </View>
    </View>
  </View>,
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
        "marginBottom": 10,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#000000",
            "fontFamily": "System",
            "fontWeight": "400",
          },
          {
            "textAlign": "left",
          },
          [
            {
              "textAlign": "left",
            },
            {
              "color": "rgba(0, 0, 0, 0.87)",
              "fontFamily": "System",
              "fontWeight": "400",
              "writingDirection": "ltr",
            },
            [
              {
                "fontSize": 14,
                "letterSpacing": 0.25,
                "lineHeight": 20,
                "marginVertical": 2,
              },
              {
                "color": "rgb(28, 28, 30)",
                "fontSize": 16,
                "textAlign": "center",
              },
            ],
          ],
        ]
      }
    >
      – 
      18:00
    </Text>
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#6cc3d5",
          "borderColor": "transparent",
          "borderRadius": 4,
          "borderStyle": "solid",
          "borderWidth": 0,
          "color": "#e8e8e8",
          "elevation": 0,
          "marginLeft": 5,
          "minWidth": "auto",
          "shadowColor": "#000000",
          "shadowOffset": {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0,
          "shadowRadius": 0,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "disabled": true,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "overflow": "hidden",
            },
            {
              "borderRadius": 4,
            },
          ]
        }
        testID="Xanax_hour2"
      >
        <View
          style={
            [
              {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
              },
              {
                "flexDirection": "row-reverse",
              },
            ]
          }
        >
          <View
            style={
              {
                "marginLeft": -4,
                "marginRight": 12,
              }
            }
          >
            <Text />
          </View>
          <Text
            numberOfLines={1}
            selectable={false}
            style={
              [
                {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                {
                  "textAlign": "left",
                },
                [
                  {
                    "letterSpacing": 1,
                    "marginHorizontal": 16,
                    "marginVertical": 9,
                    "textAlign": "center",
                  },
                  {
                    "marginHorizontal": 8,
                  },
                  {
                    "textTransform": "uppercase",
                  },
                  {
                    "color": "rgba(0, 0, 0, 0.32)",
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  {
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  undefined,
                ],
              ]
            }
          >
            Już zażyłem/-am
          </Text>
        </View>
      </View>
    </View>
  </View>,
]
`;
